### DISCUSSION TITLE
Making sure I'm on the latest version for arduino-esp32

### DISCUSSION THREAD

**Discussion by Petermarcu** [espressif/arduino-esp32#8325] [Q&A] [answered]
I'm having an issue where I'm not getting good reads from the shiftIn function but if I copy the implmentation of shiftIn into my code everything works. I'm wondering if I somehow don't have the latest implementation of shiftIn. What do I need to check to make sure I have it? I am using the Arduino IDE on Windows 11.

**Comment by bertmelis** [comment, marked as answer]
You can update in the boards manager window in Arduino IDE.

**Comment by Petermarcu** [comment]
Thanks. I had done that. I am wondering if there is an actual bug here then. I don't think I should be able to replace the shiftIn library function with a local copy from github and have it address the problem unless the library code is out of date.

**Reply by bertmelis** [reply]
What version are you on? Double check the URL in the boards manager.

**Reply by Petermarcu** [reply]
I'm on 2.09 of the esp32 package. 

If I copy the contents of the shiftIn implementation and undo the following [commit](https://github.com/espressif/arduino-esp32/commit/c0345eafbfa4c7f5d4b045e318feef0576af9fe2), then my shiftInLocal function behaves as expected and now behaves the way my code on an Arduino does.

Here is the working local function that I now have: 

```
uint8_t shiftInLocal(uint8_t dataPin, uint8_t clockPin, uint8_t bitOrder) {
    uint8_t value = 0;
    uint8_t i;

    for(i = 0; i < 8; ++i) {
        digitalWrite(clockPin, HIGH);
        if(bitOrder == LSBFIRST)
            value |= digitalRead(dataPin) << i;
        else
            value |= digitalRead(dataPin) << (7 - i);
        digitalWrite(clockPin, LOW);
    }
    return value;
}
```

**Reply by Petermarcu** [reply]
I figured out that I had a rogue clock pulse causing my issue.
